### PR TITLE
NEW: Add extension hook for field-specific validation

### PR DESCRIPTION
**Pull request by lozcalver**
See also:

- https://github.com/silverstripe/silverstripe-asset-admin/pull/1305
- https://github.com/silverstripe/silverstripe-tagfield/pull/234
- https://github.com/silverstripe/silverstripe-subsites/pull/510

**Rationale:**

As far as I’m aware, these are currently the only ways you can perform custom validation on form fields:

1. Subclass a form field and override `validate()`
2. Write a custom `Validator` and use that in place of `RequiredFields`
3. Perform validation in your form handler method(s)

These approaches all “work”, but they’re far from optimal. Adding a hook like this allows for cleaner project code, plus makes it easier to distribute validation functionality in modules.

**Example:**

This is a (slightly daft) example to illustrate the basic functionality this would offer:

```php
// EmailFieldValidationExtension.php
class EmailFieldValidationExtension extends Extension
{
    protected $shouldRejectPlusAddresses = false;

    public function setRejectPlusAddresses($reject)
    {
        $this->shouldRejectPlusAddresses = $reject;
        return $this->owner;
    }

    public function updateValidationResult(&$result, $validator)
    {
        if ($this->shouldRejectPlusAddresses && strpos($this->owner->Value(), '+') !== false) {
            $result = false;
            $validator->validationError(
                $this->owner->getName(),
                'Sorry, we’re evil and don’t allow + aliases.'
            );
        }
    }
}

// SomeControllerWithAForm.php
$fields = FieldList::create(
    EmailField::create('Email', 'Where we should send spam to')
        ->setRejectPlusAddresses(true)
)
```

### CODE REVIEW

File: src/Forms/FormField.php
Comment (lozcalver) [inline src/Forms/FormField.php:1233]: This is broken out into a separate method to save a bunch of duplication across FormField subclasses like:

```php
$result = true;
$this->extend('updateValidationResult', $result, $validator);
return $result;
```

For some fields (like `ConfirmedPasswordField`) this would end up being copy-pasted many times in one method, so it felt better to break it out like this.

The only other alternative I could spot would be to refactor all the `validate()` methods to take this approach:

```php
$result = true;
if (failedValidation()) {
    $result = false;
}
$this->extend('updateValidationResult', $result, $validator);
return $result;
```

That works for simple fields (I essentially did that for `TextField` for example) but for `ConfirmedPasswordField` it means a big refactor and removing a bunch of early-returns, which just felt like a lot of work for no real gain in readability.

File: src/Forms/NumericField.php
Comment (michalkleiner) [inline src/Forms/NumericField.php:205]: for consistency with MultiSelectField change set the $result first

```suggestion
            $result = false;
            $validator->validationError(
                $this->name,
                _t(
                    'SilverStripe\\Forms\\NumericField.VALIDATION',
                    "'{value}' is not a number, only numbers can be accepted for this field",
                    ['value' => $this->originalValue]
                )
            );
```

**Comment by GuySartorelli** [comment]
This looks good... though I do have a small concern that it will be expected for this to "just work" for form fields, but then we'll implement a new field and forget to do this, or people will expect this to work for 3rd party fields which haven't done this, which leads to an inconsistent DX

**Comment by michalkleiner** [comment]
Even though it was touching a lot of files I was also sort of ok with the change, thinking about the same thing as Guy, i.e. how can we make sure this works most of the times across all custom fields and 3rd party modules and custom validators etc.
I guess we can only do so much. Updating the docs around form field validation suggesting adding the hook to custom fields. Perhaps something like controllers check if custom init method makes sure parent init is called via onBeforeInit or something along those lines?

**Comment by lozcalver** [comment]
Yeah I understand that, it’s kinda like using `beforeUpdateCMSFields()` in that it _should_ be used by all core code/modules but there’s nothing to enforce/encourage that beside the docs. I’m not sure we can really do it programatically without:

A) abusing the deprecation API;
B) causing BC issues by triggering warnings in a minor release, or;
C) delaying this until 5.0 and then triggering warnings

Open to suggestions 🙂

**Comment by lozcalver** [comment]
Have changed the target branch to 5.0 with a view to implementing option C above, but I don’t think there’s a way to do it that really makes sense.

As field implementations don’t necessarily call `parent::validate()`, any check for “has called extendValidationResult” would have to live outside of `FormField`, which means adding it to specific validator classes like `RequiredFields`, so we’re just moving the problem instead of really fixing it because a custom `Validator` may not perform these changes. It would also get quite messy for composite fields (e.g. `MoneyField` in core) that call `validate()` on their child fields, as `validate()` would have to perform this validation _as well as_ the top-level validator.

I think it makes more sense to just leave this. If a developer doesn’t call `->extendValidationResult()`, it’s not the end of the world - it’s a fairly simple addition to add to a module, and this is a new feature so there’s no BC risk.

**Review by michalkleiner — APPROVED** [review]
The reasoning makes sense to me, it's a ✅ from me but another pair of eyes would be good. There should be a corresponding docs PR linked, too, describing the addition and how to use it.

**Comment by lozcalver** [comment]
Docs PR: https://github.com/silverstripe/developer-docs/pull/144

**Comment by lozcalver** [comment]
Bump

**Comment by michalkleiner** [comment]
I've approved this already but would prefer also @GuySartorelli's or @maxime-rainville's once over so we could get this to `5`. It would need a separate entry into the changelogs both for the beta and the unreleased stable.

**Comment by GuySartorelli** [comment]
Sorry, I keep meaning to come back to this.

On the balance I'm okay with this - but can you please include appropriate unit tests and update documentation (both the changelog and if we have validation docs there too) to tell people that this is a new best practice?

I'll retarget this to `5.0` so it can be included in the stable 5.0.0 release.

**Comment by lozcalver** [comment]
Thanks @GuySartorelli, I’ve added a new test with a few assertions to cover this and have updated the [docs pr](https://github.com/silverstripe/developer-docs/pull/144) to add a note to the 5.0.0 changelog, linking through to the validation docs which contain more detail

**Comment by michalkleiner** [comment]
Thanks @kinglozzer for taking the time and effort to get this polished and ready to be merged in, we do appreciate it! 👍

File: src/Forms/CompositeField.php
Comment (GuySartorelli) [inline src/Forms/CompositeField.php]:
```suggestion
        return $this->extendValidationResult($valid, $validator);
```
`$valid` isn't passed by reference to that method - plus we should do them all the same way

File: src/Forms/FormField.php
Comment (GuySartorelli) [reply src/Forms/FormField.php:1233]: This is a great solution - makes it really easy for third-party formfields to use this functionality.

File: tests/php/Forms/FormFieldTest.php
Comment (GuySartorelli) [inline tests/php/Forms/FormFieldTest.php:481]: Is there an easy way to loop through `ClassInfo::subclassesFor(FormField::class)` and run through this test with all our formfields? That would be a good way to ensure:
1, we haven't missed any this time around (there will be some in other modules we'll have to come back and add this to, for example, and running this test against the sink would catch them)
2. any new form fields have this functionality applied

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/CompositeField.php
Comment (lozcalver) [reply src/Forms/CompositeField.php]: Good spot!

File: tests/php/Forms/FormFieldTest.php
Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php:481]: I’ve been trying to achieve this but I can’t do it nicely. A lot of the subclasses have their own constructor argument requirements, and while I can hardcode specific arguments for specific tests, it’s going to make the test quite fragile (and long!). This is the best I could come up with, do you think it’s worth adding?

```php
    public function testValidationExtensionHooksAreCalledOnFormFieldSubclasses()
    {
        // Can't use a provider for this due to class manifest weirdness
        $formFieldClasses = ClassInfo::subclassesFor(FormField::class, false);
        foreach ($formFieldClasses as $formFieldClass) {
            // Skip over anything GridField-related
            if (strpos($formFieldClass, 'GridField') !== false) {
                continue;
            }

            $reflection = new ReflectionClass($formFieldClass);
            // Skip abstract classes like MultiSelectField
            if ($reflection->isAbstract()) {
                continue;
            }

            /** @var FormField $field */
            $field = $reflection->newInstanceWithoutConstructor();
            // Skip readonly and data-less fields
            if (!$field->hasData() || $field->isReadonly()) {
                continue;
            }

            $args = ['Test'];
            if ($formFieldClass === NullableField::class) {
                $args = [TextField::create('Test2')];
            } else if ($formFieldClass === TreeDropdownField::class) {
                $args = ['Test', 'Test', Group::class];
            } else if ($formFieldClass === PermissionCheckboxSetField::class) {
                $args = ['Test', 'Test', Permission::class, 'Test'];
            } else if ($formFieldClass === SelectionGroup::class) {
                $args = ['Test', []];
            }

            $mock = $this->getMockBuilder($formFieldClass)
                ->setConstructorArgs($args)
                ->setMethods(['extendValidationResult'])
                ->getMock();

            $mock->expects($this->once())
                ->method('extendValidationResult')
                ->will($this->returnValue(true));

            $this->assertTrue($mock->validate(new RequiredFields()));
        }
    }
```

Comment (GuySartorelli) [reply tests/php/Forms/FormFieldTest.php:481]: > // Skip over anything GridField-related

Does this not work with gridfields?

> // Skip readonly and data-less fields

I'm guessing `validate()` doesn't get called on these? That would make sense, since the user can't set values for them.

> This is the best I could come up with, do you think it’s worth adding?

Yes, definitely. This will catch any scenario where we add new fields but forget about the new validation extending. Worst case we need to add another "else if" (maybe make it a switch but I'm fine with it as is for now) when adding a new field. It'll also catch if we break this with any given field.

Well, worst case we do something really funky and the new field doesn't work with this, but we can cross that bridge if we come to it.

Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php:481]: >> // Skip over anything GridField-related

>Does this not work with gridfields?

Yes, but you can’t really “validate” a GridField anyway. The main reason for skipping was to reduce the number of "elseif" blocks, but I’ll include them for now and we can see how long it is and whether we want to exclude any

>> // Skip readonly and data-less fields

> I'm guessing validate() doesn't get called on these? That would make sense, since the user can't set values for them.

It does, again I just skipped them to reduce the number of elseif blocks. As above I’ll add everything for now and see how long it ends up

Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php:481]: Okay I’ve pushed that test, it didn’t end up as long as I’d feared 😄

**Comment by GuySartorelli** [comment]
Looks great, thank you!
Looks like there's just one failure here to sort out.

**Comment by lozcalver** [comment]
Weird, that doesn’t happen locally so perhaps another module being loaded for CI runs has a formfield I missed (proving that this test is necessary 😅). Annoyingly I can’t spot a way to add the current class being tested to any error messages that PHPUnit outputs, so I’ve hacked in some temporary debugging output

File: tests/php/Forms/FormFieldTest.php
Comment (GuySartorelli) [inline tests/php/Forms/FormFieldTest.php]:
```suggestion
                ->onlyMethods(['extendValidationResult'])
```
Seems `setMethods()` is deprecated.

File: tests/php/Forms/FormFieldTest.php
Comment (GuySartorelli) [inline tests/php/Forms/FormFieldTest.php]:
```suggestion
            try {
                $isValid = $mock->validate(new RequiredFields());
            } catch (ExpectationFailedException $e) {
                $this->fail("Expectation failed for '$formFieldClass' class: {$e->getMessage()}");
            }
            $this->assertTrue($isValid, "$formFieldClass should be valid");
```
This should make it clear which field was being tested when the `$mock->expects($this->once())` expectation fails - and added message for the `assertTrue()` as well for good measure.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Slightly more robust and permanent debugging suggestion

File: tests/php/Forms/FormFieldTest.php
Comment (GuySartorelli) [reply tests/php/Forms/FormFieldTest.php]: Obviously also need to add `use PHPUnit\Framework\ExpectationFailedException;`

**Comment by GuySartorelli** [comment]
Looks like the failure was from a test formfield.... may make sense to add a `if (is_a($formFieldClass, TestOnly::class, true)) {continue;}` in there (maybe just add to the existing condition which `continue`s) - we don't care if formfields that were created for testing _other things_ fail this test.

File: tests/php/Forms/FormFieldTest.php
Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php]: This doesn’t work 😢 looks like mock expectations aren’t checked until after the entire test has run

Comment (GuySartorelli) [reply tests/php/Forms/FormFieldTest.php]: Oh, weird. It worked for me locally when I changed `$this->once()` to `$this->never()`...

Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php]: That’s bizarre - it does work for `->never()`, but not for `once()` (I’m testing making the `once()` expectation fail by manually removing the `->extendValidationResult()` call from `EmailField`)...

Comment (GuySartorelli) [reply tests/php/Forms/FormFieldTest.php]: Bizarre indeed.... We might just not be able to add the message for that expectation then. 

This might be worth trying: https://stackoverflow.com/a/56677638 explicitly invoking an assertion for it, if that's still valid API.

But not the end of the world if we can't get a nice message in there. At least if it fails again in the future we can add a temporary debug line in like you did this time to see what class failed it.

Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php]: Looks like that API has been removed. Calling this method (`$mock->__phpunit_verify())` works, but it’s not officially supported API so it’s risky:

https://github.com/sebastianbergmann/phpunit/blob/ddf1b957bd90342cecdaec56f62e98d410187cd9/src/Framework/MockObject/Api/Api.php#L72-L79

Ideally we’d do this with a dataProvider, as that does seem to include the class name in the error, but dataProviders are fetched _so_ early that our manifests aren’t ready, and even `get_declared_classes()` won’t work without manually querying autoloaders and including files... nasty.

Comment (lozcalver) [reply tests/php/Forms/FormFieldTest.php]: Oh wait, I’ve found something else that’ll work:

```php
$mock = $this->getMockBuilder($formFieldClass)
    ->setConstructorArgs($args)
    ->onlyMethods(['extendValidationResult'])
    ->getMock();

$mock->expects($invocationRule = $this->once())
    ->method('extendValidationResult')
    ->will($this->returnValue(true));

$isValid = $mock->validate(new RequiredFields());
$this->assertTrue($isValid, "$formFieldClass should be valid");

try {
    $invocationRule->verify();
} catch (\Exception $e) {
    $this->fail("Expectation failed for '$formFieldClass' class: {$e->getMessage()}");
}
```

File: tests/php/Forms/FormFieldTest.php
Comment (lozcalver) [inline tests/php/Forms/FormFieldTest.php:611]: This API isn’t covered by phpunit’s BC promise, but it hasn’t changed from 8.x to 10.x: https://github.com/sebastianbergmann/phpunit/blob/10.0.9/src/Framework/MockObject/Rule/InvokedCount.php. I’ve included the comment for info so if it causes failures in future, this block can just be removed and the test will still run fine, just with less debugging info on error.

**Comment by lozcalver** [comment]
It should be okay now - the remaining failure for `UploadField` is because this depends on https://github.com/silverstripe/silverstripe-asset-admin/pull/1305. At least it proves the extra debugging code works as expected 😛

**Comment by GuySartorelli** [comment]
Rebased on top of 2.0 to get rid of the unrelated CI failures.

**Comment by GuySartorelli** [comment]
Nice. I've just rebased https://github.com/silverstripe/silverstripe-asset-admin/pull/1305 and once that's green I'll merge it and then this rerun CI here (just in case there's some other PR further down the line that _also_ needs to be updated).

We also run framework tests against kitchen sink - have you had a look at whether the new test passes with all of kitchen sink installed? I'm guessing there are a few fields there that will need updating as well.

**Comment by lozcalver** [comment]
> We also run framework tests against kitchen sink - have you had a look at whether the new test passes with all of kitchen sink installed? I'm guessing there are a few fields there that will need updating as well.

No it doesn’t pass, and after looking at it I’m beginning to wonder though if it’s a good idea to iterate over _all_ subclasses of `FormField` like this.

Form fields for other modules might have their own argument requirements, even if they don’t need any changes following this PR they still have to be added to that `switch` statement. It feels wrong to be including a bunch of classes from other modules framework tests like this.

An example is `Symbiote\AdvancedWorkflow\FormFields\WorkflowField::__construct()` - even though it doesn’t have overload validation, it still needs a `case` statement added for it because its arguments are different, and those arguments include a `WorkflowDefinition` typed argument which means even more “external” code in core tests

**Comment by GuySartorelli** [comment]
Yeah that does seem like a bit of an anti-pattern. But we will need a test like that _somewhere_ in kitchen sink. We don't really have a  repository that exists to hold unit tests like that.... I guess we could add that to frameworktest and configure kitchen sink to include unit tests from frameworktest...
@emteknetnz this seems like the sort of thing you'd probably have thoughts about how to solve. Do you have any thoughts on how to handle this?

**Comment by emteknetnz** [comment]
Use class_exists() within module for a class that exists in a non-framework module? We already do it quite a bit

https://github.com/silverstripe/silverstripe-framework/blob/5/tests/php/Security/MemberTest.php#L1785
```php
    public function testMapInCMSGroupsNoLeftAndMain()
    {
        if (class_exists(LeftAndMain::class)) {
            $this->markTestSkipped('LeftAndMain must not exist for this test.');
        }
        $result = Member::mapInCMSGroups();
        $this->assertInstanceOf(Map::class, $result);

        $this->assertEmpty($result, 'Without LeftAndMain, no groups are CMS groups.');
    }
```

We install silverstripe/installer for every github actions CI run, so the classes you're after will exist a lot of the time?

Is that a valid way to solve this?

**Comment by GuySartorelli** [comment]
The problem isn't that classes might not exist - the problem is that this test loops through _all_ formfields, and when we have kitchen sink installed that means the framework unit test needs to include information about how to instantiate all of the kitchen sink formfields that have their own constructors.

It may be that there's only a few and it doesn't make the test particularly messy. It may be that it makes the test quite a bit messier than it currently is. But mostly it means framework's unit test ends up having cases for a bunch of very-not-core classes.

I think it's _probably_ okay, given that it's a unit test so it won't be affecting anyone's project. But it's also not the ideal scenario.

**Comment by emteknetnz** [comment]
While it's messy i think it's much cleaner than introducing an entirely new paradigm of putting "shared" unit tests in separate module (frameworktest)

**Comment by GuySartorelli** [comment]
I agree with that on the balance. And I also think this test _is_ necessary.
@kinglozzer could you please add the necessary changes for this to pass in kitchen sink?

**Comment by lozcalver** [comment]
I’ve pushed a commit which covers everything in the kitchensink, I’ve raised a couple of new PRs for modules that did need this to be added (https://github.com/silverstripe/silverstripe-tagfield/pull/234, https://github.com/silverstripe/silverstripe-subsites/pull/510).

**Comment by GuySartorelli** [comment]
Thank you!
I've merged the asset-admin PR and rerun the failed CI jobs in this PR.
Merge on green - the other PRs will follow.

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**
